File: trunk/Source/Assets/Scripts/custom/AudioManager.js
#pragma strict
#pragma downcast

import System.Collections.Generic;
import System.IO;

public class AudioManager
{
	private static var Instance : AudioManager = null;
    private function AudioManager() {;} 												// private Constructor protected
    public static function Get(): AudioManager 										// Singleton Instance Accessor 
    {
    	if(Instance == null)
    		Instance = new AudioManager();
        return Instance;
    }


    public function Play( clip : AudioClip, emitter : Transform) : AudioSource
    {
        return Play(clip, emitter, 1f, 1f);
    }
 
    public function Play( clip : AudioClip, emitter : Transform, volume : float) : AudioSource
    {
        return Play(clip, emitter, volume, 1f);
    }

    public function Play( clip : AudioClip, emitter : Transform, volume : float, pitch : float) : AudioSource
    {
        //Create an empty game object
        var go : GameObject = new GameObject( "Audio: " +  clip.name);
        go.transform.position = emitter.position;
        go.transform.parent = emitter;
 
        //Create the source
        var source : AudioSource = go.AddComponent.<AudioSource>() as AudioSource;
        source.clip = clip;
        source.volume = volume;
        source.pitch = pitch;
        source.Play();
        GameObject.Destroy( go, clip.length);
        return source;
    }
 
   
    public function PlayLoop( clip : AudioClip, emitter : Transform, volume : float, pitch : float) : AudioSource
    {
        //Create an empty game object
        var go : GameObject = new GameObject( "Audio: " +  clip.name);
        go.transform.position = emitter.position;
        go.transform.parent = emitter;
 
        //Create the source
        var source : AudioSource = go.AddComponent.<AudioSource>() as AudioSource;
        source.clip = clip;
        source.volume = volume;
        source.pitch = pitch;
        source.loop = true;
        source.Play();
//        GameObject.Destroy( go, clip.length);
        return source;
    }  
    /// <summary>
    /// Plays a sound by creating an empty game object with an AudioSource
    /// and attaching it to the given transform (so it moves with the transform). Destroys it after it finished playing.
    /// </summary>
    /// <param name="clip"></param>
    /// <param name="emitter"></param>
    /// <param name="volume"></param>
    /// <param name="pitch"></param>
    /// <returns></returns>   
    
 
    public function Play( clip : AudioClip, point : Vector3) : AudioSource
    {
        return Play(clip, point, 1f, 1f);
    }
 
    public function Play( clip : AudioClip,  point : Vector3, volume : float) : AudioSource
    {
        return Play(clip, point, volume, 1f);
    }
 
    public function Play( clip : AudioClip,  point : Vector3,  volume : float, pitch : float) : AudioSource
    {
        //Create an empty game object
        var go : GameObject = new GameObject("Audio: " + clip.name);
        go.transform.position = point;
 
        //Create the source
        var source : AudioSource = go.AddComponent(AudioSource);
        source.clip = clip;
        source.volume = volume;
        source.pitch = pitch;
        source.Play();
        GameObject.Destroy( go, clip.length);
        return source;
    }
    
    /// <summary>
    /// Plays a sound at the given point in space by creating an empty game object with an AudioSource
    /// in that place and destroys it after it finished playing.
    /// </summary>
    /// <param name="clip"></param>
    /// <param name="point"></param>
    /// <param name="volume"></param>
    /// <param name="pitch"></param>
    /// <returns></returns> 
    
          
    
//    function PlaySound ( soundName : AudioClip, soundDelay : float)
//	{
//		if ( !audio.isPlaying && Time.time > soundRate )
//		{
//			soundRate = Time.time + soundDelay;
//			audio.clip = soundName;
//			audio.Play();
//			yield WaitForSeconds ( audio.clip.length );
//		}
//	}



function StartMusic( soundTrack : AudioClip, emitter : Transform )
{
	var music = Play( soundTrack, emitter);
	

	while(!MusicStoped)
		yield PlayMusic( music );
}
private var MusicStoped : boolean = false;

function StopMusic()
{
	MusicStoped = true;
}

function PlayMusic( soundTrack : AudioSource )
{

	if (!soundTrack.isPlaying && soundTrack.clip.isReadyToPlay)
	{
	    soundTrack.Play();
	}
	else
	{
	    Debug.Log("waiting - isplaying : " + soundTrack.isPlaying + " isreadyToPlay : " + soundTrack.clip.isReadyToPlay);
	}
	yield;
}

};

File: trunk/Source/Assets/Scripts/Managers/AudioManager.cs

using UnityEngine;
using System.Collections;
 
public class AudioManager : MonoBehaviour
{
	public bool SoundEnable = true;
	public bool _MusicEnable = true;
	public AudioSource Music;
	
	public bool MusicEnable
	{
		get
		{
			return this._MusicEnable;
		}
		set
		{
			if (value == this._MusicEnable)
				return;
			this._MusicEnable = value;
			if (this._MusicEnable)
				Managers.Audio.ResumeMusic();
			else
				Managers.Audio.PauseMusic();
		}
	}

	void Awake()
	{
        if (Camera.main != null)
        {
            //Music = gameObject.AddComponent<AudioSource>();
            Music =(Camera.main.gameObject).AddComponent<AudioSource>();
            Music.playOnAwake = false;
            Music.Stop();
            Music.loop = true;
        }
	}
	
    public AudioSource Play(AudioClip clip, Transform emitter)
    {
        return Play(clip, emitter, 1f, 1f);
    }
 
    public AudioSource Play(AudioClip clip, Transform emitter, float volume)
    {
        return Play(clip, emitter, volume, 1f);
    }
 
    /// <summary>
    /// Plays a sound by creating an empty game object with an AudioSource
    /// and attaching it to the given transform (so it moves with the transform). Destroys it after it finished playing.
    /// </summary>
    /// <param name="clip"></param>
    /// <param name="emitter"></param>
    /// <param name="volume"></param>
    /// <param name="pitch"></param>
    /// <returns></returns>
	public AudioSource Play(AudioClip clip, Transform emitter, float volume, float pitch)
	{
		if (!SoundEnable)
			return null;

		GameObject go = new GameObject("Audio: " + clip.name);
		go.transform.position = emitter.position;
		go.transform.parent = emitter;
		
		AudioSource audioSrc = go.AddComponent<AudioSource>();
		audioSrc.clip = clip;
		audioSrc.volume = volume;
		audioSrc.pitch = pitch;
		audioSrc.Play();
		Object.Destroy( go, clip.length);
		return audioSrc;
	}
 
    public AudioSource Play(AudioClip clip, Vector3 point)
    {
        return Play(clip, point, 1f, 1f);
    }
 
    public AudioSource Play(AudioClip clip, Vector3 point, float volume)
    {
        return Play(clip, point, volume, 1f);
    }
 
    /// <summary>
    /// Plays a sound at the given point in space by creating an empty game object with an AudioSource
    /// in that place and destroys it after it finished playing.
    /// </summary>
    /// <param name="clip"></param>
    /// <param name="point"></param>
    /// <param name="volume"></param>
    /// <param name="pitch"></param>
    /// <returns></returns>
    public AudioSource Play(AudioClip clip, Vector3 point, float volume, float pitch)
    {
        if (!SoundEnable)
            return null;

        //Create an empty game object
        GameObject go = new GameObject("Audio: " + clip.name);
        go.transform.position = point;
 
        //Create the source
        AudioSource source = go.AddComponent<AudioSource>();
        source.clip = clip;
        source.volume = volume;
        source.pitch = pitch;
        source.Play();
        Destroy(go, clip.length);
        return source;
    }

	public AudioSource PlayMusic(AudioClip clip, float volume = 1, float pitch =1)
    {
		if (!MusicEnable)
			return null;
		
		if (Music.clip == clip)
			return Music;
		
		if (Music.isPlaying)
			Music.Stop();
		
		Music.clip = clip;
		Music.volume = volume;
		Music.pitch = pitch;
		Music.Play();
		
		return Music;
    }
	
	public void PauseMusic()
	{
		if (!Music.isPlaying)
			return;
		this.Music.Pause();
	}
	
	public void ResumeMusic()
	{
		if (Music.isPlaying)
			return;
		this.Music.Play();
	}
	
	public AudioSource StopMusic()
    {
		if ( Music.isPlaying )
			Music.Stop();
		
        return Music;
    }

    public void CheckAllSound( bool active)
    {
		if (active)
			return;

		foreach (AudioSource audioSource in Object.FindObjectsOfType(typeof (AudioSource)))
		{
			if (audioSource != Music)
				Object.Destroy(audioSource);
		}
    }

    //public void CheckAllSound(bool FxActive, bool MusicActive)
    //{
    //    if (!MusicActive)
    //        PauseMusic();
    //    else
    //        ResumeMusic();

    //    if (!FxActive)
    //        foreach (AudioSource sonidos in GameObject.FindObjectsOfTypeAll(typeof(AudioSource)))
    //            if (sonidos != Music)
    //                Destroy(sonidos);
    //}
}